sw/device/silicon_creator/rom_ext/rescue.c

```diff
@@ -32,32 +35,37 @@ rom_error_t flash_firmware_block(rescue_state_t *state) {
         .write = kMultiBitBool4True,
         .erase = kMultiBitBool4True,
     });
-    for (uint32_t addr = state->flash_start; addr <= state->flash_limit;
+    for (uint32_t addr = state->flash_start; addr < state->flash_limit;
          addr += kFlashPageSize) {
       HARDENED_RETURN_IF_ERROR(
           flash_ctrl_data_erase(addr, kFlashCtrlEraseTypePage));
     }
     state->flash_offset = state->flash_start;
   }
-  if (state->flash_offset <= state->flash_limit) {
+  if (state->flash_offset < state->flash_limit) {
     HARDENED_RETURN_IF_ERROR(flash_ctrl_data_write(
         state->flash_offset, sizeof(state->data) / sizeof(uint32_t),
         state->data));
     state->flash_offset += sizeof(state->data);
   } else {
+    xmodem_cancel(iohandle);
     return kErrorRescueImageTooBig;
   }
   return kErrorOk;
 }
 
-rom_error_t flash_owner_block(rescue_state_t *state) {
-  // FIXME: validate that we're in a state capable of accepting an owner
-  // block and validate the block before flashing it.
-  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(&kFlashCtrlInfoPageOwnerSlot1,
-                                                 kFlashCtrlEraseTypePage));
-  HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
-      &kFlashCtrlInfoPageOwnerSlot1, 0, sizeof(state->data) / sizeof(uint32_t),
-      state->data));
+rom_error_t flash_owner_block(rescue_state_t *state, boot_data_t *bootdata) {
+  if (bootdata->ownership_state == kOwnershipStateUnlockedAny ||
+      bootdata->ownership_state == kOwnershipStateLockedUpdate ||
+      bootdata->ownership_state == kOwnershipStateUnlockedEndorsed) {
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_info_erase(
+        &kFlashCtrlInfoPageOwnerSlot1, kFlashCtrlEraseTypePage));
+    HARDENED_RETURN_IF_ERROR(flash_ctrl_info_write(
+        &kFlashCtrlInfoPageOwnerSlot1, 0,
+        sizeof(state->data) / sizeof(uint32_t), state->data));
+  } else {
+    dbg_printf("error: cannot accept owner_block in current state\r\n");
+  }
```

Comment:
I think it would be good to have a bit of extra hardening here given that faulting this condition would result in a vulnerability. @johannheyszl can calculate the risk potential.

Adding redundant checks before performing flash erase seem a bit complicated from an FI perspective. What if we try to set a flag after performing the flash write, and then before returning at the end of the function perform a redundant check based on the state of the flag. 

```
hardened_bool_t updated_flash = kHardenedBoolFalse;

if (critical block) {
  // flash erase & write
  updated_flash = kHardenedBoolFalse;
}
...

if (launder32(updated_flash) == kHardenedBoolFalse) {
  // read back flash check.
  // clear flash if it doesn't match.
}
return kErrorOk;
```

Reply:
I'm game for hardening this, but we should understand if its really needed.

Uploading the owner config here is only a staging operation.  If an attacker is able to defeat the `if` condition and upload a configuration to page 1, it still has to be validated on next boot.

The potential ownership states are either `LockedOwner` or `Recovery` (if the chip were in one of the unlocked states, there'd be no reason to try to defeat this `if` statement).
- In `LockedOwner`, the owner page has to be validated by [`locked_owner_init`](https://github.com/lowRISC/opentitan/blob/earlgrey_es_sival/sw/device/silicon_creator/lib/ownership/ownership.c#L57-L98).  That function requires one of the two pages to be sealed to be valid, and if one of the pages is invalid, it erases the invalid page and copies the valid page there.  If neither page is sealed, the ownership state transitions to `Recovery`.
- In `Recovery` the chip refuses to honor the contents of the owner pages.

_If_ the owner can thunk the `ownership_state` in `ownership_init` on the next boot to one of the unlocked states, _then_ owner page 1 will be honored (if it is signed).  We should probably harden the switch/case in `ownership_init`.

Reply:
Hi @cfrantz, this explanation makes sense. I agree with your assessment.